barcode_utils.py

```diff
@@ -0,0 +1,166 @@
+
+    encrypted_string = _barcode_raw_encrypt(data_to_encrypt, key=key)
+
+    # check to make sure it worked.
+    decrypted = get_badge_num_from_barcode(encrypted_string, salt, key)
+    if decrypted['badge_num'] != badge_num or decrypted['event_id'] != event_id:
+        raise ValueError("didn't encode correctly")
+
+    # check to make sure this barcode number is valid for Code 128 barcode
+    verify_barcode_is_valid_code128(encrypted_string)
+
+    return encrypted_string
+
+
+def get_badge_num_from_barcode(barcode_num, salt=None, key=None):
+    salt = config['secret']['barcode_salt'] if not salt else salt
+    key = bytes(config['secret']['barcode_key'],'ascii') if not key else key
+
+    decrypted = _barcode_raw_decrypt(barcode_num, key=key)
+
+    result = dict()
+
+    result['event_id'] = struct.unpack('>B', bytearray([decrypted[0]]))[0]
+
+    badge_bytes = bytearray(bytes([0, decrypted[1], decrypted[2], decrypted[3]]))
+    result['badge_num'] = struct.unpack('>I', badge_bytes)[0] - salt
+
+    return result
+
+
+def verify_barcode_is_valid_code128(encrypted_string):
+    for c in encrypted_string:
+        if c not in code128._charset_b:
+            raise ValueError("contains a char not valid in a code128 barcode")
+
+
+def _barcode_raw_encrypt(value, key):
+    if len(value) != 4:
+        raise ValueError("invalid barcode input: needs to be exactly 4 bytes")
+
+    # skip32 generates 4 bytes output from 4 bytes input
+    _encrypt = True
+    skip32.skip32(key, value, _encrypt)
+
+    # raw bytes aren't suitable for a Code 128 barcode though,
+    # so convert it to base58 encoding
+    # which is just some alphanumeric and numeric chars and is
+    # designed to be vaguely human.  this takes our 4 bytes and turns it into 6 chars
+    encrypted_value = base64.encodebytes(value).decode('ascii')
+
+    # important note: because we are not an even multiple of 3 bytes, base64 needs to pad
+    # the resulting string with equals signs.  we can strip them out knowing that our length is 4 bytes
+    # IF YOU CHANGE THE LENGTH OF THE ENCRYPTED DATA FROM 4 BYTES, THIS WILL NO LONGER WORK.
+    encrypted_value = encrypted_value.replace('==\n', '')
+
+    if len(encrypted_value) != 6:
+        raise ValueError("Barcode encoding failure: result should be 6 characters")
+
+    return encrypted_value
+
+
+def _barcode_raw_decrypt(value, key):
+    # raw bytes aren't suitable for a Code 128 barcode though,
+    # so convert it to base64 encoding
+    # which is just some alphanumeric and numeric chars and is
+    # designed to be vaguely human.  this takes our 4 bytes and turns it into 6ish bytes
+
+    # important note: because we are not an even multiple of 3 bytes, base64 needs to pad
+    # the resulting string with equals signs.  we can strip them out knowing that our length is 4 bytes
+    # IF YOU CHANGE THE LENGTH OF THE ENCRYPTED DATA FROM 4 BYTES, THIS WILL NO LONGER WORK.
+
+    if len(value) != 6:
```

Comment:
The error message probably shouldn't be exactly the same between two functions - even though you're checking the same thing, you'd want to indicate that the error here is the value being read in, not imply that something was encoded.